From 654771cd602208a3147be2e49136b1fff6a19547 Mon Sep 17 00:00:00 2001
From: Sriram R <srirrama@codeaurora.org>
Date: Mon, 16 Nov 2020 10:32:39 -0800
Subject: [PATCH] sigma-dut: Add MBSSID support

Add HE MBSSID commands support in sigma-dut and fill corresponding
hostapd conf.

NonTxBSSIndex
NumNonTxBSS
in set_wireless, get_mac_address and set_security commands

Currently the max no of mbss vaps is set to 2. This
can be extended by changing the macro value in future
based on test case requirements.

Support is available for SAE (Hash-to-Element) in 6GHz and WPA2-PSK in
5GHz/2.45GHz.

Signed-off-by: Sriram R <srirrama@codeaurora.org>
Signed-off-by: Aloka Dixit <alokad@codeaurora.org>
---
 ap.c        | 619 +++++++++++++++++++++++++++++++++++-----------------
 sigma_dut.h |  60 +++--
 2 files changed, 460 insertions(+), 219 deletions(-)

--- a/ap.c
+++ b/ap.c
@@ -647,6 +647,36 @@ static enum sigma_cmd_result cmd_ap_set_
 	if (val)
 		dut->program = sigma_program_to_enum(val);
 
+	val = get_param(cmd, "NonTxBSSIndex");
+	if (val) {
+		int non_tx_bss_idx;
+
+		if (dut->ap_mbssid != VALUE_ENABLED) {
+			send_resp(dut, conn, SIGMA_ERROR,
+				  "errorCode,MBSSID Not enabled");
+			return STATUS_SENT_ERROR;
+		}
+
+		non_tx_bss_idx = atoi(val);
+
+		if (non_tx_bss_idx > dut->mbss.num_non_tx_bss) {
+			send_resp(dut, conn, SIGMA_ERROR,
+				  "errorCode,Invalid NonTxBSSIndex");
+			return STATUS_SENT_ERROR;
+		}
+
+		val = get_param(cmd, "SSID");
+		if (val) {
+			if (strlen(val) > MAX_SSID_LEN)
+				return INVALID_SEND_STATUS;
+
+			snprintf(dut->mbss.conf[non_tx_bss_idx - 1].ssid,
+				 MAX_SSID_LEN, "%s", val);
+		}
+
+		return SUCCESS_SEND_STATUS;
+	}
+
 	val = get_param(cmd, "WLAN_TAG");
 	if (val) {
 		wlan_tag = atoi(val);
@@ -1954,6 +1984,27 @@ static enum sigma_cmd_result cmd_ap_set_
 		}
 	}
 
+	val = get_param(cmd, "NumNonTxBSS");
+	if (val) {
+		int num_non_tx_bss;
+
+		if (dut->ap_mbssid != VALUE_ENABLED) {
+			send_resp(dut, conn, SIGMA_ERROR,
+				  "errorCode,MBSSID Not enabled");
+			return STATUS_SENT_ERROR;
+		}
+
+		num_non_tx_bss = atoi(val);
+
+		if (num_non_tx_bss > MAX_NON_TX_VAP_COUNT) {
+			send_resp(dut, conn, SIGMA_ERROR,
+				  "errorCode,MBSS Vap Count Support Exceeded");
+			return STATUS_SENT_ERROR;
+		}
+
+		dut->mbss.num_non_tx_bss = num_non_tx_bss;
+	}
+
 	val = get_param(cmd, "TWT_RespSupport");
 	if (val) {
 		if (strcasecmp(val, "enable") == 0) {
@@ -2269,6 +2320,56 @@ static enum sigma_cmd_result cmd_ap_set_
 		}
 	}
 
+	val = get_param(cmd, "NonTxBSSIndex");
+	if (val) {
+		int non_tx_bss_idx;
+
+		if (dut->ap_mbssid != VALUE_ENABLED) {
+			send_resp(dut, conn, SIGMA_ERROR,
+				  "errorCode,MBSSID Not enabled");
+			return STATUS_SENT_ERROR;
+		}
+
+		non_tx_bss_idx = atoi(val);
+
+		if (non_tx_bss_idx > dut->mbss.num_non_tx_bss) {
+			send_resp(dut, conn, SIGMA_ERROR,
+				  "errorCode,Invalid NonTxBSSIndex");
+			return STATUS_SENT_ERROR;
+		}
+
+		val = get_param(cmd, "KEYMGNT");
+		if (!val)
+			val = get_param(cmd,"KeyMgmtType");
+		if (val) {
+			if (strcasecmp(val, "SAE") == 0) {
+				dut->mbss.conf[non_tx_bss_idx - 1].key_mgmt = AP_WPA2_SAE;
+			} else {
+				if (dut->use_6g) {
+					send_resp(dut, conn, SIGMA_INVALID,
+						  "errorCode,Unsupported 6GHz KeyMgmtType");
+					return 0;
+				}
+				else if (strcasecmp(val, "WPA2-PSK") == 0 ||
+					 (security && strcasecmp(security, "PSK") == 0 &&
+					  strcasecmp(val, "WPA2") == 0)) {
+					dut->mbss.conf[non_tx_bss_idx - 1].key_mgmt = AP_WPA2_PSK;
+				}
+			}
+		}
+
+		val = get_param(cmd, "PSK");
+		if (val) {
+			if (strlen(val) > MAX_PSK_LEN - 1)
+				return INVALID_SEND_STATUS;
+
+			snprintf(dut->mbss.conf[non_tx_bss_idx - 1].psk, MAX_PSK_LEN,
+				 "%s", val);
+		}
+
+		return SUCCESS_SEND_STATUS;
+	}
+
 	val = get_param(cmd, "KEYMGNT");
 	if (!val)
 		val = get_param(cmd,"KeyMgmtType");
@@ -7995,6 +8096,177 @@ static int fill_dual_radio_caps(struct s
 
 }
 
+void ath11k_fw_test_commands(struct sigma_dut *dut, const char *ifname)
+{
+	int pdev_id;
+
+	if (strncmp(ifname, "wlan0", 5) == 0) {
+		pdev_id = 1;
+	} else if (strncmp(ifname, "wlan1", 5) == 0) {
+		pdev_id = 2;
+	} else {
+		pdev_id = 3;
+	}
+
+	/* disable sending basic triggers */
+	run_system_wrapper(dut, "ath11k-fwtest -i %s -m 0x47 -v 0 42 0",
+			   ifname);
+	/* disable MU BAR */
+	run_system_wrapper(dut, "ath11k-fwtest -i %s -m 0x47 -v 0 64 1",
+			   ifname);
+	/* disable PSD Boost*/
+	run_system_wrapper(dut, "ath11k-fwtest -i %s -m 0x48 -v 0 142 1",
+			   ifname);
+	/* Enable mix bw */
+	run_system_wrapper(dut, "ath11k-fwtest -i %s -m 0x47 -v 0 141 1",
+			   ifname);
+	/* Disable preferred AC */
+	run_system_wrapper(dut, "ath11k-fwtest -i %s -m 0x48 -v 0 186 0",
+			   ifname);
+	if (dut->ap_he_ulofdma == VALUE_ENABLED) {
+		/* he_dynamic_muedca 0 */
+		run_system_wrapper(dut,
+				   "ath11k-fwtest -t 2 -i %s -m 0x00 -p %d 0xBF 0x00",
+				   ifname, pdev_id);
+	}
+
+	/* enable full_band_probing */
+	run_system_wrapper(dut, "ath11k-fwtest -i %s -m 0x47 -v 0 194 0",
+			   ifname);
+
+	if (dut->ap_he_ulofdma == VALUE_ENABLED) {
+		/* Disable DL OFDMA */
+		if (dut->ap_he_dlofdma != VALUE_ENABLED) {
+			run_system_wrapper(dut,
+					   "ath11k-fwtest -t 1 -i %s -m 0x00 -v 0 32770 111",
+					   ifname);
+		}
+	}
+
+	if (dut->ap_txBF) {
+		/* Ignore TBTT for NDP */
+		run_system_wrapper(dut,
+				   "ath11k-fwtest -i %s -m 0x48 -v 0 2 1",
+				   ifname);
+		/* cv query enable */
+		run_system_wrapper(dut,
+				   "ath11k-fwtest -i %s -m 0x47 -v 0 7 1",
+				   ifname);
+		/* override TPC calculations & set TxBF flag to true */
+		run_system_wrapper(dut,
+				   "ath11k-fwtest -i %s -m 0x47 -v 0 47 1",
+				   ifname);
+	}
+	if (dut->ap_he_mimo == MIMO_DL) {
+		mac80211_set_mu_brp_commands(dut, ifname);
+		/* Increase MU traffic */
+		run_system_wrapper(dut,
+				   "ath11k-fwtest -i %s -m 0x48 -v 0 100 2",
+				   ifname);
+	}
+
+	if (dut->ap_he_ulofdma == VALUE_ENABLED) {
+		/* he_mu_edca 1 */
+		run_system_wrapper(dut,
+				   "ath11k-fwtest -t 1 -i %s -m 0x00 -v 0 0x8F 0x01",
+				   ifname);
+		/* Disable sounding for UL OFDMA */
+		run_system_wrapper(dut, "ath11k-fwtest -i %s -m 0x47 -v 0 7 0",
+				   ifname);
+		/* enable the equal RU allocation */
+		run_system_wrapper(dut,
+				   "ath11k-fwtest -i %s -m 0x4b -v 0 0 1",
+				   ifname);
+		/* he ul ltf */
+		run_system_wrapper(dut,
+				   "ath11k-fwtest -t 1 -i %s -m 0x00 -v 0 0x8D 0x03",
+				   ifname);
+		/*he ul gi */
+		run_system_wrapper(dut,
+				   "ath11k-fwtest -t 1 -i %s -m 0x00 -v 0 0x89 0x03",
+				   ifname);
+		/* Trigger Interval */
+		run_system_wrapper(dut,
+				   "ath11k-fwtest -t 2 -i %s -m 0x00 -p %d 0xA2 0x02",
+				   ifname, pdev_id);
+		/* disable enable_ul_ofdma_efficiency_check */
+		run_system_wrapper(dut,
+				   "ath11k-fwtest -i %s -m 0x47 -v 0 131 0",
+				   ifname);
+		/* Disable ul_mumimo_efficiency_check */
+		run_system_wrapper(dut,
+				   "ath11k-fwtest -i %s -m 0x47 -v 0 204 0",
+				   ifname);
+		/* Set random RU allocation */
+		run_system_wrapper(dut, "ath11k-fwtest -i %s -m 0x47 -v 0 9 1",
+				   ifname);
+		/* To set TBTT PPDU duration (us) */
+		run_system_wrapper(dut, "ath11k-fwtest -i %s -m 0x48 -v 0 63 1908",
+				   ifname);
+
+		/* enable twt related commands by default, during basic trigger setting for
+		 * ofdma this can be reset */
+		/* BSR Trigger disable/enable UL trigger frame scheduling */
+		run_system_wrapper(dut, "ath11k-fwtest -i %s -m 0x47 -v 0 42 7",
+				   ifname);
+
+		/* Set twt trigger type to basic trigger */
+		run_system_wrapper(dut,
+				   "ath11k-fwtest -i %s -m 0x47 -v 0 95 6",
+				   ifname);
+
+		/* Configure hw stuck ssr threshold */
+		run_system_wrapper(dut,
+				   "ath11k-fwtest -i %s -m 0x47 -v 0 43 8",
+				   ifname);
+
+		/* UL trigger duration */
+		run_system_wrapper(dut,
+				   "ath11k-fwtest -i %s -m 0x48 -v 0 63 1000",
+				   ifname);
+	}
+
+	if (dut->ap_he_ppdu == PPDU_MU &&
+	    dut->ap_he_dlofdma == VALUE_ENABLED) {
+		/* Increase the min TX time limit for MU MIMO to disable MU MIMO
+		 * scheduling */
+		run_system_wrapper(dut,
+				   "ath11k-fwtest -i %s -m 0x47 -v 0 11 1000000",
+				   ifname);
+		/* Increase the max TX time limit for DL OFDMA to enable OFDMA
+		 * scheduling */
+		run_system_wrapper(dut,
+				   "ath11k-fwtest -i %s -m 0x47 -v 0 17 1000000",
+				   ifname);
+		/* Disable 'force SU schedule' to enable MU sch */
+		run_system_wrapper(dut,"ath11k-fwtest -i %s -m 0x47 -v 0 8 0",
+				   ifname);
+		/* Enable MU 11ax support in sch algo */
+		run_system_wrapper(dut,
+				   "ath11k-fwtest -i %s -m 0x47 -v 0 29 0",
+				   ifname);
+		/* Enable to sort RU allocation */
+		run_system_wrapper(dut, "ath11k-fwtest -i %s -m 0x4b -v 0 2 1",
+				   ifname);
+	}
+
+	if (dut->ap_bcc == VALUE_ENABLED && dut->ap_fixed_rate) {
+		if (dut->use_6g)
+			run_system_wrapper(dut, "iw %s set bitrates he-mcs-6 1:%d\n",
+					   ifname, dut->he_mcs);
+		else if (dut->ap_channel <= 11)
+			run_system_wrapper(dut, "iw %s set bitrates he-mcs-2.4 1:%d\n",
+					   ifname, dut->he_mcs);
+		else
+			run_system_wrapper(dut, "iw %s set bitrates he-mcs-5 1:%d\n",
+					   ifname, dut->he_mcs);
+	}
+
+	if (dut->he_sounding == VALUE_ENABLED)
+		run_system_wrapper(dut, "ath11k-fwtest -i %s -m 0x47 -v 0 7 0",
+				   ifname);
+}
+
 enum sigma_cmd_result cmd_ap_config_commit(struct sigma_dut *dut,
 					   struct sigma_conn *conn,
 					   struct sigma_cmd *cmd)
@@ -9213,6 +9485,91 @@ skip_key_mgmt:
 		fprintf(f, "oce=%d\n",
 			dut->dev_role == DEVROLE_STA_CFON ? 2 : 1);
 	}
+
+	if (dut->program == PROGRAM_HE && dut->ap_mbssid == VALUE_ENABLED) {
+		uint8_t num_bss, bssid_mask, max_bss_count;
+		uint8_t init_val, base_val;
+		uint8_t max_bssid_ind = 0;
+		unsigned char bssid[6];
+		char ifname2[50];
+		int i;
+
+
+		if (get_hwaddr(ifname, bssid)) {
+			fclose(f);
+			return -2;
+		}
+
+		if (!dut->mbss.num_non_tx_bss)
+			return -2;
+
+		/* Sanity check to indicate below logic holds good
+		 * for upto 256 bss(including tx,non-tx bss) support in sigma-dut
+		 * Can be extended if really required
+		 */
+		if (dut->mbss.num_non_tx_bss > 255)
+			return -2;
+
+		num_bss = dut->mbss.num_non_tx_bss;
+
+		do {
+			max_bssid_ind++;
+			num_bss = num_bss >> 1;
+		} while (num_bss);
+
+		max_bss_count = 1 << max_bssid_ind;
+		bssid_mask = max_bss_count - 1;
+
+		base_val = bssid[5] & ~bssid_mask;
+		init_val = bssid[5] & bssid_mask;
+
+		fprintf(f, "\nmultiple_bssid=1\n");
+
+		for (i = 0; i < dut->mbss.num_non_tx_bss; i++) {
+			fprintf(f, "\n");
+
+			/* start from 1 for non tx bssid */
+			bssid[5] = base_val + ((init_val + i + 1) % max_bss_count);
+
+			snprintf(ifname2, sizeof(ifname2), "%s-%d", ifname, i+1);
+			fprintf(f, "bss=%s\n", ifname2);
+			if (sigma_hapd_ctrl)
+				fprintf(f, "ctrl_interface=%s\n", sigma_hapd_ctrl);
+			else
+				fprintf(f, "ctrl_interface=/var/run/hostapd\n");
+			fprintf(f, "ssid=%s\n", dut->mbss.conf[i].ssid);
+			if (dut->bridge)
+				fprintf(f, "bridge=%s\n", dut->bridge);
+
+			if (dut->ap_wme)
+				fprintf(f, "wmm_enabled=1\n");
+
+			if (dut->ap_wmmps == AP_WMMPS_ON)
+				fprintf(f, "uapsd_advertisement_enabled=1\n");
+
+			fprintf(f, "bssid=%02x:%02x:%02x:%02x:%02x:%02x\n",
+				bssid[0], bssid[1], bssid[2], bssid[3],
+				bssid[4], bssid[5]);
+
+			if (dut->mbss.conf[i].key_mgmt == AP_WPA2_SAE) {
+				fprintf(f, "wpa=2\n");
+				fprintf(f, "wpa_key_mgmt=SAE\n");
+				fprintf(f, "wpa_pairwise=CCMP\n");
+				fprintf(f, "ieee80211w=2\n");
+				fprintf(f, "sae_password=%s\n", dut->mbss.conf[i].psk);
+				if (dut->use_6g)
+					fprintf(f, "sae_pwe=1\n");
+			}
+
+			if (dut->mbss.conf[i].key_mgmt == AP_WPA2_PSK) {
+				fprintf(f, "wpa=2\n");
+				fprintf(f, "wpa_key_mgmt=WPA-PSK\n");
+				fprintf(f, "wpa_pairwise=CCMP\n");
+				fprintf(f, "wpa_passphrase=%s\n", dut->mbss.conf[i].psk);
+			}
+		}
+	}
+
 	fclose(f);
 	if (dut->use_hostapd_pid_file)
 		kill_hostapd_process_pid(dut);
@@ -9510,15 +9867,7 @@ skip_key_mgmt:
 	}
 
 	if(drv == DRIVER_MAC80211 && dut->program == PROGRAM_HE) {
-		int pdev_id;
-
-		if (strcmp(ifname, "wlan0") == 0) {
-			pdev_id = 1;
-		} else if (strcmp(ifname, "wlan1") == 0) {
-			pdev_id = 2;
-		} else {
-			pdev_id = 3;
-		}
+		int i;
 
 		/* A WAR done for MU_BAR test case. We are disabling anything
 		   related to MU MIMO when dl_ofdma test case is running
@@ -9530,148 +9879,12 @@ skip_key_mgmt:
 			dut->ap_mu_txBF = 0;
 		}
 
-		/* disbale sending basic triggers */
-		run_system_wrapper(dut, "ath11k-fwtest -i %s -m 0x47 -v 0 42 0",
-				   ifname);
-		/* disable MU BAR */
-		run_system_wrapper(dut, "ath11k-fwtest -i %s -m 0x47 -v 0 64 1",
-			           ifname);
-		/* disable PSD Boost*/
-		run_system_wrapper(dut, "ath11k-fwtest -i %s -m 0x48 -v 0 142 1",
-				   ifname);
-		/* Enable mix bw */
-		run_system_wrapper(dut, "ath11k-fwtest -i %s -m 0x47 -v 0 141 1",
-				   ifname);
-		/* Disable preferred AC */
-		run_system_wrapper(dut, "ath11k-fwtest -i %s -m 0x48 -v 0 186 0",
-				   ifname);
-		if (dut->ap_he_ulofdma == VALUE_ENABLED) {
-			/* he_dynamic_muedca 0 */
-			run_system_wrapper(dut,
-					   "ath11k-fwtest -t 2 -i %s -m 0x00 -p %d 0xBF 0x00",
-					   ifname, pdev_id);
-		}
-
-		/* enable full_band_probing */
-		run_system_wrapper(dut, "ath11k-fwtest -i %s -m 0x47 -v 0 194 0",
-				   ifname);
-
-		if (dut->ap_he_ulofdma == VALUE_ENABLED) {
-			/* Disable DL OFDMA */
-			if (dut->ap_he_dlofdma != VALUE_ENABLED) {
-				run_system_wrapper(dut,
-						   "ath11k-fwtest -t 1 -i %s -m 0x00 -v 0 32770 111",
-						   ifname);
-			}
-		}
-
-		if (dut->ap_txBF) {
-			/* Ignore TBTT for NDP */
-			run_system_wrapper(dut,
-					   "ath11k-fwtest -i %s -m 0x48 -v 0 2 1",
-					   ifname);
-		        /* cv query enable */
-			run_system_wrapper(dut,
-					   "ath11k-fwtest -i %s -m 0x47 -v 0 7 1",
-					   ifname);
-			/* override TPC calculations & set TxBF flag to true */
-			run_system_wrapper(dut,
-					   "ath11k-fwtest -i %s -m 0x47 -v 0 47 1",
-					   ifname);
-		}
-		 if (dut->ap_he_mimo == MIMO_DL) {
-			mac80211_set_mu_brp_commands(dut, ifname);
-			/* Increase MU traffic */
-			run_system_wrapper(dut,
-					   "ath11k-fwtest -i %s -m 0x48 -v 0 100 2",
-					   ifname);
-		}
-
-		if (dut->ap_he_ulofdma == VALUE_ENABLED) {
-			/* he_mu_edca 1 */
-			run_system_wrapper(dut,
-					   "ath11k-fwtest -t 1 -i %s -m 0x00 -v 0 0x8F 0x01",
-					   ifname);
-			/* Disable sounding for UL OFDMA */
-			run_system_wrapper(dut, "ath11k-fwtest -i %s -m 0x47 -v 0 7 0",
-					   ifname);
-			/* enable the equal RU allocation */
-			run_system_wrapper(dut,
-					   "ath11k-fwtest -i %s -m 0x4b -v 0 0 1",
-					   ifname);
-			/* he ul ltf */
-			run_system_wrapper(dut,
-					   "ath11k-fwtest -t 1 -i %s -m 0x00 -v 0 0x8D 0x03",
-					   ifname);
-			/*he ul gi */
-			run_system_wrapper(dut,
-					   "ath11k-fwtest -t 1 -i %s -m 0x00 -v 0 0x89 0x03",
-					   ifname);
-			/* Trigger Interval */
-			run_system_wrapper(dut,
-					   "ath11k-fwtest -t 2 -i %s -m 0x00 -p %d 0xA2 0x02",
-					   ifname, pdev_id);
-			/* disable enable_ul_ofdma_efficiency_check */
-			run_system_wrapper(dut,
-					   "ath11k-fwtest -i %s -m 0x47 -v 0 131 0",
-					   ifname);
-			/* Disable ul_mumimo_efficiency_check */
-			run_system_wrapper(dut,
-					   "ath11k-fwtest -i %s -m 0x47 -v 0 204 0",
-					   ifname);
-			/* Set random RU allocation */
-			run_system_wrapper(dut, "ath11k-fwtest -i %s -m 0x47 -v 0 9 1",
-					   ifname);
-			/* To set TBTT PPDU duration (us) */
-			run_system_wrapper(dut, "ath11k-fwtest -i %s -m 0x48 -v 0 63 1908",
-					   ifname);
-
-			/* enable twt related commands by default, during basic trigger setting for
-			 * ofdma this can be reset */
-
-			/* BSR Trigger disable/enable UL trigger frame scheduling */
-			run_system_wrapper(dut, "ath11k-fwtest -i %s -m 0x47 -v 0 42 7",
-					   ifname);
-
-			/* Set twt trigger type to basic trigger */
-			run_system_wrapper(dut,
-					   "ath11k-fwtest -i %s -m 0x47 -v 0 95 6",
-					   ifname);
-
-			/* Configure hw stuck ssr threshold */
-			run_system_wrapper(dut,
-					   "ath11k-fwtest -i %s -m 0x47 -v 0 43 8",
-					   ifname);
-
-			/* UL trigger duration */
-			run_system_wrapper(dut,
-					   "ath11k-fwtest -i %s -m 0x48 -v 0 63 1000",
-					   ifname);
-		}
-
-		if (dut->ap_he_ppdu == PPDU_MU &&
-		    dut->ap_he_dlofdma == VALUE_ENABLED) {
-			/* Increase the min TX time limit for MU MIMO to disable MU MIMO
-			 * scheduling */
-			run_system_wrapper(dut,
-					   "ath11k-fwtest -i %s -m 0x47 -v 0 11 1000000",
-					   ifname);
-			/* Increase the max TX time limit for DL OFDMA to enable OFDMA
-			 * scheduling */
-			run_system_wrapper(dut,
-					   "ath11k-fwtest -i %s -m 0x47 -v 0 17 1000000",
-					   ifname);
-			/* Disable 'force SU schedule' to enable MU sch */
-			run_system_wrapper(dut,"ath11k-fwtest -i %s -m 0x47 -v 0 8 0",
-				           ifname);
-			/* Enable MU 11ax support in sch algo */
-			run_system_wrapper(dut,
-					   "ath11k-fwtest -i %s -m 0x47 -v 0 29 0",
-					   ifname);
-			/* Enable to sort RU allocation */
-			run_system_wrapper(dut, "ath11k-fwtest -i %s -m 0x4b -v 0 2 1",
-					   ifname);
-		}
+		ath11k_fw_test_commands(dut, ifname);
+		for (i = 0; i < dut->mbss.num_non_tx_bss; i++) {
+			char ifname2[50];
+			snprintf(ifname2, sizeof(ifname2), "%s-%d", ifname, i+1);
+			ath11k_fw_test_commands(dut, ifname2);
+ 		}
 	}
 
 	if (dut->program == PROGRAM_60GHZ) {
@@ -9705,23 +9918,6 @@ skip_key_mgmt:
 				  "errorCode,Fail to set wps_version_number");
 			return STATUS_SENT;
 		}
-
-		if (dut->ap_bcc == VALUE_ENABLED && dut->ap_fixed_rate) {
-			if (dut->use_6g)
-				run_system_wrapper(dut, "iw %s set bitrates he-mcs-6 1:%d\n",
-						   ifname, dut->he_mcs);
-			else if (dut->ap_channel <= 11)
-				run_system_wrapper(dut, "iw %s set bitrates he-mcs-2.4 1:%d\n",
-						   ifname, dut->he_mcs);
-			else
-				run_system_wrapper(dut, "iw %s set bitrates he-mcs-5 1:%d\n",
-						   ifname, dut->he_mcs);
-		}
-
-		if (dut->he_sounding == VALUE_ENABLED)
-			run_system_wrapper(dut, "ath11k-fwtest -i %s -m 0x47 -v 0 7 0",
-					   ifname);
-
 	}
 
 	dut->hostapd_running = 1;
@@ -10310,6 +10506,8 @@ static enum sigma_cmd_result cmd_ap_rese
 		}
 		if (get_openwrt_driver_type() == OPENWRT_DRIVER_ATHEROS)
 			dut->ap_dfs_mode = AP_DFS_MODE_ENABLED;
+
+		memset(&dut->mbss, 0, sizeof(struct mbss_support));
 	}
 
 	dut->ap_oper_chn = 0;
@@ -11844,6 +12042,28 @@ static enum sigma_cmd_result cmd_ap_get_
 
 	get_if_name(dut, ifname, sizeof(ifname), wlan_tag);
 
+	val = get_param(cmd, "NonTxBSSIndex");
+	if (val) {
+		int non_tx_bss_idx;
+		char idx[4];
+
+		if (dut->ap_mbssid != VALUE_ENABLED) {
+			send_resp(dut, conn, SIGMA_ERROR,
+				  "errorCode,MBSSID Not enabled");
+			return STATUS_SENT_ERROR;
+		}
+
+		non_tx_bss_idx = atoi(val);
+
+		if (non_tx_bss_idx > dut->mbss.num_non_tx_bss) {
+			send_resp(dut, conn, SIGMA_ERROR,
+				  "errorCode,Invalid NonTxBSSIndex");
+			return STATUS_SENT_ERROR;
+		}
+		snprintf(idx, sizeof(idx), "-%d", non_tx_bss_idx);
+		strlcat(ifname, idx, sizeof(ifname));
+	}
+
 	s = socket(AF_INET, SOCK_DGRAM, 0);
 	if (s < 0)
 		return -1;
--- a/sigma_dut.h
+++ b/sigma_dut.h
@@ -354,6 +354,44 @@ enum akm_suite_values {
 
 };
 
+enum ap_key_mgmt {
+	AP_OPEN,
+	AP_WPA2_PSK,
+	AP_WPA_PSK,
+	AP_WPA2_EAP,
+	AP_WPA_EAP,
+	AP_WPA2_EAP_MIXED,
+	AP_WPA2_PSK_MIXED,
+	AP_WPA2_SAE,
+	AP_WPA2_PSK_SAE,
+	AP_SUITEB,
+	AP_WPA2_OWE,
+	AP_WPA2_EAP_OSEN,
+	AP_WPA2_FT_EAP,
+	AP_WPA2_FT_PSK,
+	AP_WPA2_EAP_SHA256,
+	AP_WPA2_PSK_SHA256,
+	AP_WPA2_ENT_FT_EAP,
+	AP_OSEN,
+};
+
+/* Update this macro for supporting multiple mbss vaps */
+#define MAX_NON_TX_VAP_COUNT	2
+#define MAX_SSID_LEN		15
+#define MAX_PSK_LEN		20
+
+struct mbss_config {
+	int non_tx_bss_idx;
+	char ssid[MAX_SSID_LEN];
+	enum ap_key_mgmt key_mgmt;
+	char psk[MAX_PSK_LEN];
+};
+
+struct mbss_support {
+	int num_non_tx_bss;
+	struct mbss_config conf[MAX_NON_TX_VAP_COUNT];
+};
+
 struct sigma_dut {
 	const char *main_ifname;
 	char *main_ifname_2g;
@@ -505,26 +543,7 @@ struct sigma_dut {
 	enum value_not_set_enabled_disabled ap_dyn_bw_sig;
 	int ap_sgi80;
 	int ap_p2p_mgmt;
-	enum ap_key_mgmt {
-		AP_OPEN,
-		AP_WPA2_PSK,
-		AP_WPA_PSK,
-		AP_WPA2_EAP,
-		AP_WPA_EAP,
-		AP_WPA2_EAP_MIXED,
-		AP_WPA2_PSK_MIXED,
-		AP_WPA2_SAE,
-		AP_WPA2_PSK_SAE,
-		AP_SUITEB,
-		AP_WPA2_OWE,
-		AP_WPA2_EAP_OSEN,
-		AP_WPA2_FT_EAP,
-		AP_WPA2_FT_PSK,
-		AP_WPA2_EAP_SHA256,
-		AP_WPA2_PSK_SHA256,
-		AP_WPA2_ENT_FT_EAP,
-		AP_OSEN,
-	} ap_key_mgmt;
+	enum ap_key_mgmt ap_key_mgmt;
 	enum ap_tag_key_mgmt {
 		AP2_OPEN,
 		AP2_OSEN,
@@ -981,6 +1000,7 @@ struct sigma_dut {
 	int ap_6g_only;
 	int ap_6g_unsol_prb_resp;
 	enum if_band ap_dual_band;
+	struct mbss_support mbss;
 	int ocvc;
 	int client_privacy;
 	int saquery_oci_freq;
